cpp/Ice/forwarder/Forwarder.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "Forwarder.h"
+
+using namespace std;
+
+ForwardingServer::Forwarder::Forwarder(Ice::ObjectPrx targetTemplate) : _targetTemplate{std::move(targetTemplate)} {}
+
+void
+ForwardingServer::Forwarder::dispatch(Ice::IncomingRequest& request, function<void(Ice::OutgoingResponse)> sendResponse)
+{
+    // Make a copy of the current object carried by the request.
+    const Ice::Current current{request.current()};
+
+    // Create a proxy with the the desired identity and facet.
+    Ice::ObjectPrx target = _targetTemplate.ice_identity(current.id).ice_facet(current.facet);
+
+    if (current.requestId == 0)
+    {
+        // The incoming request is one-way, so we reconfigure target to be one-way.
+        target = target.ice_oneway();
+    }
+
+    // Read the encapsulation containing the in-parameters.
+    const std::byte* inEncapsulationStart = nullptr;
+    int32_t inEncapsulationSize = 0;
+    request.inputStream().readEncapsulation(inEncapsulationStart, inEncapsulationSize);
+
+    // Make the invocation asynchronously. This call reports most exceptions through its exception callback.
+    target.ice_invokeAsync(
+        current.operation,
+        current.mode,
+        make_pair(inEncapsulationStart, inEncapsulationStart + inEncapsulationSize),
```

Comment:
This is safe because the encapsulation for the input parameters is synchronously copied to OutgoingAsync. We don’t have a way to avoid this copy, right?

Reply:
Yes, the "marshaling" is always synchronous for Ice invocations, including ice_invokeAsync.

And indeed, the marshaling - even of bytes - always allocates a new buffer. There is no public API to transfer the ownership of the in-params InputStream to the in-params OutputStream.

For colloc, we have an _internal_ API that does the opposite: "converts" an OutputStream into an InputStream. 
https://github.com/zeroc-ice/ice/blob/4f7ba6a20a54c43c54891fc74494843107d78cd4/cpp/src/Ice/CollocatedRequestHandler.cpp#L136

https://github.com/zeroc-ice/ice/blob/4f7ba6a20a54c43c54891fc74494843107d78cd4/cpp/include/Ice/InputStream.h#L103

Note that adopt is false, most likely for retries.